### ISSUE TITLE
Differences in volume ensemble sampling

### ISSUE THREAD

**Issue by balazsfabian** [openmm/openmm#4071]
I compiled OpenMM from source following the [recommended steps](http://docs.openmm.org/latest/userguide/library/02_compiling.html) with CUDA support. I tried version v7.7.0 (git tag), ed049af1b34ca and some in between.

I simulated a small, 903 TIP3P water box using the `MonteCarloBarostat` and a pressure of 1 atmosphere. I saved the volume after every 1000 steps to a logfile.

When I compute the autocorrelation of volumes of neat water boxes (903 TIP3P waters) across the two versions, I get significantly different correlation times. To condense the differences in autocorrelation into a single number, I averaged the autocorrelation curve over the first 100 points (omitting the trivial 1 at the beginning). Using the same python script, the volumes produced by v7.7 are virtually immediately uncorrelated, while those of ed049af1b34ca show long-time correlations. Here are the versions that I tested along with the average volume and `acorr[1:100].mean()`.

| Commit    |   Volume   |   Mean of acorr.   |
| --------- | ------- | -------- |
| ed049af1b | 26.5706 |  0.24906 |
| 3c7eef3d0 | 27.4298 |  0.49866 |
| dbf25377e | 26.5994 |  0.00462 |
| ae42c911a | 26.5972 |  0.00277 |
| 48b6abc1e | 26.6088 |  0.00231 |
| v8.0rc1   | 26.6036 | -0.00026 |
| v8.0beta  | 26.5929 |  0.00117 |
| v7.7      | 26.5883 |  0.00045 |

The environment.yml:
```
name: /home/tb/bafabian/conda-envs/openmm-dev
channels:
  - conda-forge
  - bioconda
  - defaults
dependencies:
  - _libgcc_mutex=0.1=conda_forge
  - _openmp_mutex=4.5=2_gnu
  - bzip2=1.0.8=h7f98852_4
  - c-ares=1.18.1=h7f98852_0
  - ca-certificates=2022.9.24=ha878542_0
  - cmake=3.25.0=h816a3e0_0
  - cython=0.29.32=py310hd8f1fbe_1
  - doxygen=1.9.5=h583eb01_0
  - expat=2.5.0=h27087fc_0
  - keyutils=1.6.1=h166bdaf_0
  - krb5=1.19.3=h08a2579_0
  - ld_impl_linux-64=2.39=hc81fddc_0
  - libblas=3.9.0=16_linux64_openblas
  - libcblas=3.9.0=16_linux64_openblas
  - libcurl=7.86.0=h2283fc2_1
  - libedit=3.1.20191231=he28a2e2_2
  - libev=4.33=h516909a_1
  - libffi=3.4.2=h7f98852_5
  - libgcc-ng=12.2.0=h65d4601_19
  - libgfortran-ng=12.2.0=h69a702a_19
  - libgfortran5=12.2.0=h337968e_19
  - libgomp=12.2.0=h65d4601_19
  - libiconv=1.17=h166bdaf_0
  - liblapack=3.9.0=16_linux64_openblas
  - libnghttp2=1.47.0=hff17c54_1
  - libnsl=2.0.0=h7f98852_0
  - libopenblas=0.3.21=pthreads_h78a6416_3
  - libsqlite=3.40.0=h753d276_0
  - libssh2=1.10.0=hf14f497_3
  - libstdcxx-ng=12.2.0=h46fd767_19
  - libuuid=2.32.1=h7f98852_1000
  - libuv=1.44.2=h166bdaf_0
  - libzlib=1.2.13=h166bdaf_4
  - make=4.3=hd18ef5c_1
  - ncurses=6.3=h27087fc_1
  - numpy=1.23.4=py310h53a5b5f_1
  - openssl=3.0.7=h166bdaf_0
  - pcre2=10.40=hc3806b6_0
  - pip=22.3.1=pyhd8ed1ab_0
  - python=3.10.6=ha86cf86_0_cpython
  - python_abi=3.10=2_cp310
  - readline=8.1.2=h0f457ee_0
  - rhash=1.4.3=h166bdaf_0
  - setuptools=65.5.1=pyhd8ed1ab_0
  - swig=4.1.0=h5d7ef5f_4
  - tk=8.6.12=h27826a3_0
  - wheel=0.38.4=pyhd8ed1ab_0
  - xz=5.2.6=h166bdaf_0
  - zlib=1.2.13=h166bdaf_4
  - zstd=1.5.2=h6239696_4
  - pip:
    - atomsmm==0.1.0
    - mpmath==1.3.0
    - openmm==7.7.0
    - pandas==2.0.1
    - python-dateutil==2.8.2
    - pytz==2023.3
    - scipy==1.10.1
    - six==1.16.0
    - sympy==1.12rc1
    - tzdata==2023.3
prefix: /home/tb/bafabian/conda-envs/openmm-dev
```

```
$ nvidia-smi 
Mon May 15 15:07:51 2023       
+-----------------------------------------------------------------------------+
| NVIDIA-SMI 470.86       Driver Version: 470.86       CUDA Version: 11.4     |
|-------------------------------+----------------------+----------------------+
| GPU  Name        Persistence-M| Bus-Id        Disp.A | Volatile Uncorr. ECC |
| Fan  Temp  Perf  Pwr:Usage/Cap|         Memory-Usage | GPU-Util  Compute M. |
|                               |                      |               MIG M. |
|===============================+======================+======================|
|   0  NVIDIA GeForce ...  Off  | 00000000:04:00.0  On |                  N/A |
|  0%   46C    P8    11W / 200W |    719MiB /  8085MiB |      0%      Default |
|                               |                      |                  N/A |
+-------------------------------+----------------------+----------------------+
```

```
$python -c "from simtk import openmm ; print(openmm.version.version)"
7.7.0.dev-130124a

# and
$python -c "from simtk import openmm ; print(openmm.version.version)"
8.0.0.dev-ed049af
```

**Comment by RaulPPelaez** [comment]
Would you please share the script you used?

**Comment by balazsfabian** [comment]
Here is a zip file containing the script and the input files: [omm.zip](https://github.com/openmm/openmm/files/11485639/omm.zip).

**Comment by balazsfabian** [comment]
> I wonder, though, are you seeing this also with the CPU implementation? If it is something that takes a reasonable amount of time.

I can confirm that running the script with version ed049af1 on CPU gives uncorrelated volumes.

**Comment by sef43** [comment]
@peastman the barostat reject step from #4049 : 
```c++
        // Reject the step.

        context.getOwner().setPeriodicBoxVectors(box[0], box[1], box[2]);
        kernel.getAs<ApplyMonteCarloBarostatKernel>().restoreCoordinates(context);
    }
```
calls `void CudaUpdateStateDataKernel::setPeriodicBoxVectors()`  which looks like it uses the `cu.getPosCellOffsets()` from the trial move rather than the original ones: 
```c++
void CudaUpdateStateDataKernel::setPeriodicBoxVectors(ContextImpl& context, const Vec3& a, const Vec3& b, const Vec3& c ) {
    vector<CudaContext*>& contexts = cu.getPlatformData().contexts;

    // If any particles have been wrapped to the first periodic box, we need to unwrap them
    // to avoid changing their positions.

    vector<Vec3> positions;
    for (auto& offset : cu.getPosCellOffsets()) {
        if (offset.x != 0 || offset.y != 0 || offset.z != 0) {
            getPositions(context, positions);
            break;
        }
    }
    
    // Update the vectors.

    for (auto ctx : contexts)
        ctx->setPeriodicBoxVectors(a, b, c);
    if (positions.size() > 0)
        setPositions(context, positions);
}
```
is this correct?

**Comment by peastman** [comment]
It doesn't matter what happens to either the positions or cell offsets in `setPeriodicBoxVectors()`, because it's immediately followed by a call to `restoreCoordinates()` which sets both back to what they were before the step was attempted.  That was the fix in #4049.

**Comment by sef43** [comment]
> It doesn't matter what happens to either the positions or cell offsets in `setPeriodicBoxVectors()`, because it's immediately followed by a call to `restoreCoordinates()` which sets both back to what they were before the step was attempted. That was the fix in #4049.

Ok understood! However, I am not convinced the `MonteCarloBarostat` is working completely correctly. The only difference before and after the #4032 fix should be the wrapping of molecule coordinates in the output trajectories, but clearly there are more significant differences as shown in the above comments. I don't understand where else the differences could come from.

**Comment by peastman** [comment]
Agreed.  The original fix in #4034 seemed like a simple one with little chance of breaking anything.  But it introduced a subtle interaction that took a long time to track down.  #4049 seemed to fix it, but I can't guarantee there isn't still some other difference in behavior.

**Comment by sef43** [comment]
I think the fix is to swap the order in the trial move of setting box vectors and scaling positions:

https://github.com/openmm/openmm/commit/148cfb94498acad59301080189a7f5f69a513cdc

`scalePositions`  also saves the current state, but the `setPeriodicBoxVectors` sometimes unwraps and resaves them. This invalidates the saved positions and offsets.

**Comment by peastman** [comment]
I think the original version is correct.  `scalePositions()` saves the positions and box vectors so that we can restore them later if the step is rejected.  It needs to happen before `setPeriodicBoxVectors()` modifies them.  Otherwise we can't restore them to what they were before.

**Comment by peastman** [comment]
Perhaps we need to split that method up?  Maybe ApplyMonteCarloBarostatKernel should define three methods: `saveCoordinates()`, `scaleCoordinates()`, and `restoreCoodinates()`.

**Comment by sef43** [comment]
> Perhaps we need to split that method up? Maybe ApplyMonteCarloBarostatKernel should define three methods: `saveCoordinates()`, `scaleCoordinates()`, and `restoreCoodinates()`.

yes that sounds like the better way to do it

**Comment by sef43** [comment]
With further testing it appears that reverting #4034 and #4049; and just doing this order swap in https://github.com/openmm/openmm/commit/148cfb94498acad59301080189a7f5f69a513cdc  fixes the original issue #4032 .

**Comment by peastman** [comment]
It might fix that particular case, but I don't think it would full resolve the issue.  Any time you call `setPeriodicBoxVectors()`, there's a possibility it will reset the cell offsets.  If you don't record both the positions and cell offsets, it isn't possible to return things to exactly the state they were in before.

**Comment by leeping** [comment]
I recently encountered a related issue when using OpenMM 8.0.0 where a large number of water molecules in my simulations all "jumped" large distances at the same time; I had not seen this behavior in earlier versions.  In the hope that #4034 would fix the issue I observed, I built the latest commit and tried running my job with it.  

Using the latest build (yesterday's commit), I found that the barostat can cause the total energy to increase over time, and if it is not damped away quickly enough by the thermostat the system can blow up.  In version 8.0.0 and older versions this behavior does not occur.  I've attached a screenshot that shows the differences.

Let me know if this is helpful in any way, as I would need some time to make a minimal test case out of it.

<img width="425" alt="Screenshot 2023-06-08 122311" src="https://github.com/openmm/openmm/assets/1441560/052fb6e0-2049-4df1-bcea-8f459715926e">

**Comment by peastman** [comment]
I have an attempted fix in #4106.  Can you try it out?

**Comment by leeping** [comment]
I just tried it, and it seems to be giving NaN for the potential energies.  If I turn the barostat off the energies are reasonable.

[openmm_0_baro1.txt](https://github.com/openmm/openmm/files/11695024/openmm_0_baro1.txt)

**Comment by peastman** [comment]
Can you provide a script to reproduce what you're doing?

**Comment by leeping** [comment]
Done - please check your email.

**Comment by peastman** [comment]
Your script works for me with the branch from #4106.

```
12:50:01 - Progress             E.T.A  Speed (ns/day)      Time(ns)       Temp(K)       Kin(kJ)       Pot(kJ)       Ene(kJ)      Vol(nm3)    Rho(kg/m3)
12:50:01 -   0.010%          6h46m00s           3.547       0.00010     179.50811   53728.24738 -388928.59715 -335200.34977     244.39715     979.21310
12:50:01 -   0.020%          3h48m27s          28.263       0.00020     181.54064   54336.59768 -389568.59205 -335231.99437     244.08161     980.47897
12:50:01 -   0.030%          2h49m25s          28.002       0.00030     181.30502   54266.07624 -389512.74114 -335246.66490     243.76328     981.75941
12:50:02 -   0.040%          2h19m32s          28.815       0.00040     180.94244   54157.55141 -389414.36508 -335256.81367     243.76328     981.75941
```

**Comment by leeping** [comment]
It's very strange - I just pulled your newest changes and rebuilt from scratch (after wiping the build folder) and I'm still getting nans on my end.  It's possible this issue might not be directly related to your fix because I usually install from conda rather than build from source these days; however, it's still strange because I don't get nans if the barostat is turned off.  Any chance it could be hardware-related?  I'm using a GTX 1080 Ti for my local run.

```
Restarting simulation from the restart file.
13:04:55 - --== Production (1000000 steps, 1.00 ns) ==--
13:04:55 - Progress will be reported every 100 steps
13:04:55 - DCD Reporter will write to dynamics_0.dcd every 10000 steps
13:04:55 - Backing up restart.p -> restart_1.p
13:04:55 - Restart information will be written to restart.p every 10000 steps
13:04:56 - Progress             E.T.A  Speed (ns/day)      Time(ns)       Temp(K)       Kin(kJ)       Pot(kJ)       Ene(kJ)      Vol(nm3)    Rho(kg/m3)
13:04:56 -   0.010%          2h03m20s          11.677       0.00010 7893977086713762.00000 2362731934991338496.00000           nan           nan     245.85206     973.41828
13:04:57 -   0.020%          1h54m36s          13.598       0.00020 18984562520080052.00000 5682234904574058496.00000           nan           nan     241.67907     990.22595
13:04:57 -   0.030%          1h53m16s          13.010       0.00030 39217487929847656.00000 11738115036835336192.00000           nan           nan     246.70536     970.05145
```

**Comment by peastman** [comment]
Can you try building the master branch?  That will tell us whether it's related to the changes in the PR.

I tested CUDA on a RTX 4080 and OpenCL on an Apple M1 Pro.  I don't know of any reason it should depend on hardware.

**Comment by peastman** [comment]
It's because, as I frequently say, Nothing Ever Works.  It doesn't mean you were doing anything wrong.  The universe is just a strange and mysterious place. :)

Anyway, it looks like the fix was successful.  So at least that works!  Thanks for testing it.

**Comment by peastman** [comment]
Fixed by #4106.

**Comment by leeping** [comment]
Wise words. :) Thanks so much for the fix!
